Subsite preview links are broken if the subsite overview is unpublished (causing WSOD)

When a comms officer was trying to share a preview link to a subsite page that wasn't public yet, they encountered a problem for those clicking the link, as they where getting the WSOD technicalk difficulties message.

On closer inspection, the following PHP error was being triggered.
> UnexpectedValueException: Object not found in SplObjectStorage->offsetGet() (line 59 of /mnt/www/html/brightonhovemain/docroot/modules/contrib/localgov_subsites/src/Plugin/Block/SubsitesNavigationBlock.php).

This occurs as the subsitenavigationblock is building all menu links using 
`$entities = $mapper->loadAndAccessCheckEntitysForTreeNodes('node', $tree, $cache);` 
and then tries to get the subsite id using 
`$subsite_id = $entities[$ancestors[0]]->id();`.
But since the overview is unpublished, this will fail as its blank.
Wrapping an empty check around this and setting subsite_id ?? NULL if blank does seem to resolve this issue, though it's not an ideal fix.

Steps to replicate
1. Create a subsite overview, but don't publish
2. Create a subsite page that uses the unpublished overview as the parent
3. Generate a preview line for the subsite page
4. Try to use the preview link in a private browsing window.

Expected result
Subsite page is shown, albeit with only the page in the navigation menu

Actual resule
WSOD, PHP error or technical difficulities message depending on setup.

Additional, wrapping this in an empty block will also cause this error on the twig
> The website encountered an unexpected error. Please try again later.
AssertionError: $url must be a string or object of type \Drupal\Core\Url in assert() (line 246 of core/lib/Drupal/Core/Template/TwigExtension.php).
Becuase the subsite overview link will have no url.
so subsite-navigation.html.twig needs to be updated with
````
{% if item.url is not empty %}
   {{ link(item.title, item.url) }}
{% endif %}
````

--- a/src/Plugin/Block/SubsitesNavigationBlock.php
+++ b/src/Plugin/Block/SubsitesNavigationBlock.php
@@ -7,6 +7,7 @@ use Drupal\Core\Cache\CacheableMetadata;
 use Drupal\Core\Entity\EntityInterface;
 use Drupal\Core\Session\AccountInterface;
 use Drupal\Core\Template\Attribute;
+use Drupal\Node\NodeInterface;
 
 /**
  * Class SubsiteNavigationBlock.
@@ -56,8 +57,17 @@ class SubsitesNavigationBlock extends SubsitesAbstractBlockBase {
       $mapper = \Drupal::service('entity_hierarchy.entity_tree_node_mapper');
       $entities = $mapper->loadAndAccessCheckEntitysForTreeNodes('node', $tree, $cache);
       $items = $this->nestTree($tree, $ancestors, $entities);
-      $subsite_id = $entities[$ancestors[0]]->id();
-      $overview_entity = $entities[$ancestors[0]];
+
+      // Sometimes the subsite is missing if the overview is unpublished and
+      // the user is anonymous, in which case set it to NULL.
+      if (!empty($entities[$ancestors[0]]) && $entities[$ancestors[0]] instanceof NodeInterface) {
+        $subsite_id = $entities[$ancestors[0]]->id();
+        $overview_entity = $entities[$ancestors[0]];
+      }
+      else {
+        $subsite_id = NULL;
+        $overview_entity = NULL;
+      }
     }
     elseif ($subsite_entity->bundle('localgov_subsites_overview')) {
       // Subsite overview page with no children.

--- a/src/Plugin/Block/SubsitesNavigationBlock.php
+++ b/src/Plugin/Block/SubsitesNavigationBlock.php
@@ -63,10 +63,18 @@ class SubsitesNavigationBlock extends SubsitesAbstractBlockBase {
       if (!empty($entities[$ancestors[0]]) && $entities[$ancestors[0]] instanceof NodeInterface) {
         $subsite_id = $entities[$ancestors[0]]->id();
         $overview_entity = $entities[$ancestors[0]];
+        $cache->addCacheableDependency($overview_entity);
       }
       else {
         $subsite_id = NULL;
         $overview_entity = NULL;
+
+        // But we still need the parent subsite for cache purposes.
+        $entities = $mapper->loadEntitiesForTreeNodesWithoutAccessChecks('node', $tree, $cache);
+        if (!empty($entities[$ancestors[0]]) && $entities[$ancestors[0]] instanceof NodeInterface) {
+          $overview_entity_for_cache = $entities[$ancestors[0]];
+          $cache->addCacheableDependency($overview_entity_for_cache);
+        }
       }
     }
     elseif ($subsite_entity->bundle('localgov_subsites_overview')) {

--- a/templates/subsite-navigation.html.twig
+++ b/templates/subsite-navigation.html.twig
@@ -49,7 +49,9 @@
         {% if item.in_active_trail %}
           {{ item.title }}
         {% else %}
-          {{ link(item.title, item.url) }}
+          {% if item.url is not empty %}
+            {{ link(item.title, item.url) }}
+          {% endif %}
         {% endif %}
         {% if item.below %}
           {{ subsite_navigation.menu_links(item.below, attributes, menu_level + 1) }}

--- a/tests/src/Functional/SubsiteBlocksTest.php
+++ b/tests/src/Functional/SubsiteBlocksTest.php
@@ -238,6 +238,7 @@ class SubsiteBlocksTest extends BrowserTestBase {
     $subsite_overview->set('localgov_subsites_hide_menu', ['value' => 0]);
     $subsite_overview->set('status', NodeInterface::NOT_PUBLISHED);
     $subsite_overview->save();
+    drupal_flush_all_caches();
     $this->drupalGet($subsite_page1->toUrl()->toString());
     $this->assertSession()->responseContains('block-localgov-subsite-navigation');
     $this->assertSession()->responseContains($subsite_page1_title);

--- a/tests/src/Functional/SubsiteBlocksTest.php
+++ b/tests/src/Functional/SubsiteBlocksTest.php
@@ -233,6 +233,16 @@ class SubsiteBlocksTest extends BrowserTestBase {
     $this->assertSession()->responseNotContains('block-localgov-subsite-navigation');
     $this->drupalGet($subsite_page1->toUrl()->toString());
     $this->assertSession()->responseNotContains('block-localgov-subsite-navigation');
+
+    // Test the navigation is still avalible with an unpublished overview node.
+    $subsite_overview->set('localgov_subsites_hide_menu', ['value' => 0]);
+    $subsite_overview->set('status', NodeInterface::NOT_PUBLISHED);
+    $subsite_overview->save();
+    $this->drupalGet($subsite_page1->toUrl()->toString());
+    $this->assertSession()->responseContains('block-localgov-subsite-navigation');
+    $this->assertSession()->responseContains($subsite_page1_title);
+    $this->assertSession()->responseContains($subsite_page2_title);
+    $this->assertSession()->pageTextNotContains($subsite_overview_title);
   }
 
 }

